Document `@source not` and `@source inline` (#2181)
tailwindlabs/tailwindcss.com: src/docs/detecting-classes-in-source-files.mdx

This can be useful when working with monorepos where your build commands run from the root of the monorepo instead of the root of each project.

### Ignoring specific paths

Use `@source not` to ignore specific paths when scanning for class names:

```css
/* [!code filename:CSS] */
@import "tailwindcss";
/* [!code highlight:2] */

> @source not "src/components/legacy";

Reviewer: I personally like starting with `./` or `../`, gets rid of the confusion where files/folders are relative to.

```

This can be useful in projects that have multiple Tailwind stylesheets where you want to make sure each one only includes the classes each stylesheet needs.

## Safelisting specific utilities

> If you need to make sure Tailwind generates certain class names that don’t exist in your content files, you can force them to be inlined by using `@source inline()`:

Reviewer 2: Slight reordering to avoid having a comment as the last item in the paragraph. Also changed "inlined" to "generated", because that feels more natural to me when talking about how Tailwind works.

```suggestion
If you need to make sure Tailwind generates certain class names that don’t exist in your content files, use `@source inline()` to force them to be generated:
```

<CodeExampleStack>
```css
/* [!code filename:CSS] */
@import "tailwindcss";
/* [!code highlight:2] */
@source inline("underline");
```

```css
/* [!code filename:Generated CSS] */
.underline {
  text-decoration: underline;
}
```

</CodeExampleStack>

### Generating variants

You can also use `@source inline()` to generate classes with variants. For example, to generate the `underline` class with hover and focus variants, add `{hover:,focus:,}` to the source input:

<CodeExampleStack>
```css
/* [!code filename:CSS] */
@import "tailwindcss";
/* [!code highlight:2] */
@source inline("{hover:,focus:,}underline");
```

```css

> /* [!code filename:Generated CSS] */
> .underline {
>   text-decoration: underline;
> }
> @media (hover: hover) {
>   .hover\\:underline:hover {
>     text-decoration: underline;
>   }
> }
> @media (focus: focus) {
>   .focus\\:underline:focus {
>     text-decoration: underline;
>   }
> }

Reviewer: In CSS these are only escaped once:
```suggestion
/* [!code filename:Generated CSS] */
.underline {
  text-decoration: underline;
}
@media (hover: hover) {
  .hover\:underline:hover {
    text-decoration: underline;
  }
}
@media (focus: focus) {
  .focus\:underline:focus {
    text-decoration: underline;
  }
}
```

```

</CodeExampleStack>

### Generating ranges

The source input is [brace expanded](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html), so you can generate multiple classes at once. For example, to generate all the red shades with hover variants, you can add a range to the source input:

<CodeExampleStack>
```css
/* [!code filename:CSS] */
@import "tailwindcss";
/* [!code highlight:2] */
@source inline("{hover:,}bg-red-{50,{100..900..100},950}");
```

```css
/* [!code filename:Generated CSS] */
.bg-red-50 {
  background-color: var(--color-red-50);
}
.bg-red-100 {
  background-color: var(--color-red-100);
}
.bg-red-200 {
  background-color: var(--color-red-200);
}

/* ... */

.bg-red-800 {
  background-color: var(--color-red-800);
}
.bg-red-900 {
  background-color: var(--color-red-900);
}
.bg-red-950 {
  background-color: var(--color-red-950);
}

> @media (hover: hover) {
>   .hover\\:bg-red-50:hover {
>     background-color: var(--color-red-50);
>   }
> 
>   /* ... */
> 
>   .hover\\:bg-red-950:hover {
>     background-color: var(--color-red-950);
>   }

Reviewer: ```suggestion
@media (hover: hover) {
  .hover\:bg-red-50:hover {
    background-color: var(--color-red-50);
  }

  /* ... */

  .hover\:bg-red-950:hover {
    background-color: var(--color-red-950);
  }
```

}
```

</CodeExampleStack>

> This generates shades of red from 100 to 900 in increments of 100, along with the first and last shades of 50 and 950. It also adds the `hover:` variant for each of those classes.

Author: @reinink Tending towards keeping the "in increments of 100" because it explains the third argument in the input.

> This generates shades of red from 100 to 900 in increments of 100, along with the first and last shades of 50 and 950. It also adds the `hover:` variant for each of those classes.

Reviewer 2: @danhollick Yup all good, if you think it's worth having, keep it 👍

### Explicitly excluding classes

If you need to avoid specific classes from being generated you can use `@source not inline()`:

```css
/* [!code filename:CSS] */
@import "tailwindcss";
/* [!code highlight:2] */
@source not inline("{hover:,focus:,}bg-red-{50,{100..900..100},950}");
```

> This will specifically exclude the shades of red from 50 to 950, along with their hover and focus variants, from being generated even if they are detected in your source files.

Reviewer 2: Tightened up a bit.

```suggestion
This will explicitly exclude the red background utilities, along with their hover and focus variants, from being generated.
```
